glyphs-reader/src/font.rs

```diff
@@ -665,8 +744,116 @@ fn default_master_idx(raw_font: &RawFont) -> usize {
         .unwrap_or(0)
 }
 
-impl From<RawFont> for Font {
-    fn from(mut from: RawFont) -> Self {
+fn extract_axis_mappings(from: &RawFont) -> BTreeMap<String, RawUserToDesignMapping> {
+    let mut axis_mappings: BTreeMap<String, RawUserToDesignMapping> = BTreeMap::new();
+    if let Some((_, axis_map)) = custom_param(&from.other_stuff, "Axis Mappings") {
+        let Plist::Dictionary(axis_map) = axis_map.get("value").unwrap() else {
+            panic!("Incomprehensible axis map");
+        };
+        for (axis_tag, mappings) in axis_map.iter() {
+            let Plist::Dictionary(mappings) = mappings else {
+                panic!("Incomprehensible mappings");
+            };
+            for (user, design) in mappings.iter() {
+                let user: f32 = user.parse().unwrap();
+                let design = design.as_f64().unwrap() as f32;
+                axis_mappings
+                    .entry(axis_tag.clone())
+                    .or_default()
+                    .push((user.into(), design.into()));
+            }
+        }
+    }
+    axis_mappings
+}
+
+impl TryFrom<RawShape> for Shape {
+    type Error = Error;
+
+    fn try_from(mut from: RawShape) -> Result<Self, Self::Error> {
+        // TODO: handle numerous unsupported attributes
+        // See <https://github.com/schriftgestalt/GlyphsSDK/blob/Glyphs3/GlyphsFileFormat/GlyphsFileFormatv3.md#differences-between-version-2>
+
+        let shape = if let Some(Plist::String(ref_glyph_name)) = from.other_stuff.remove("ref") {
+            // only components use ref
+            let transform = if let Some(pos) = &from.pos {
+                if pos.len() != 2 {
+                    return Err(Error::StructuralError(format!("Strang pos {:?}", pos)));
+                }
+                Affine::translate(pos[0].into_inner(), pos[1].into_inner())
+            } else {
+                Affine::identity()
+            };
+            Shape::Component(Component {
+                glyph_name: ref_glyph_name,
+                transform,
+                other_stuff: from.other_stuff.clone(),
+            })
+        } else {
+            // no ref; presume it's a path
+            Shape::Path(Path {
+                closed: from.closed.unwrap_or_default(),
+                nodes: from.nodes.clone().unwrap_or_default(),
+            })
+        };
+        Ok(shape)
+    }
+}
+
+fn map_and_push_if_present<T, U>(dest: &mut Vec<T>, src: Option<Vec<U>>, map: fn(U) -> T) {
+    let Some(src) = src else {
+        return; // nop
+    };
+    src.into_iter().map(map).for_each(|v| dest.push(v));
+}
+
+fn layer_to_glyph_instance(layer: Layer) -> Result<GlyphInstance, Error> {
+    let mut shapes = Vec::new();
+
+    // Glyphs v2 uses paths and components
+    map_and_push_if_present(&mut shapes, layer.paths, Shape::Path);
+    map_and_push_if_present(&mut shapes, layer.components, Shape::Component);
```

Comment:
nice use of generics and function pointer 👍

Reply:
Confidence in Rust slowly grows